README.md
atlas_usbpix_control
====================

Controlling a T3MAPS pixel detector using USBpix board, basil and python.

Setup instructions for hardware
------------------------

- USB cable from MultiIO to computer
- 2x2 Molex from FE-I4 adapter card to 2V power supply, expecting <70 mA of
current. (Usually approximately 50 mA.)
- 4x2 Molex from FE-I4 adapter card to T3MAPS GND of DO_OUT_P(5) and
  DO_OUT_P(4).
- Connect T3MAPS power (red+black jacks) to GD and VD1 pins on FE-I4 adapter
  card (near 2x2 Molex power input). Black end = GD, white = VD1.
- Connect T3MAPS pins to FE-I4 pins in the following way. Make sure the
  connectors are oriented the correct way on the ribbon so corresponding
  colors are actually connected (and not inverted). The connections on
  T3MAPS go on the INSIDE pin of the 2-pin pair. The outside pin is for
  ground.

    FE-I4 ADAPTER CARD PIN     CABLE COLOR              T3MAPS PIN

    GNDD (near DobOutP)-------BLACK--------------------GND of GCfgDout

    GNDD (near DobOutP)-------PINK---------------------GND of PrmpVbp

    IoMxIn2-------------------BROWN--------------------SRIN_ALL

    SelCmd--------------------RED----------------------SRCK_G

    IoMxOut0------------------GREY(END OF CONNECTOR)---SR_OUT

    IoMxIn3-------------------YELLOW-------------------GCfgCK

    IoMxIn1-------------------GREEN--------------------Dacld

    IoMxSel1------------------PALE BLUE----------------Stbld

- Connect bias cable to the LEMO jack. It should be a -4V bias.

Testing the connections
-----------------------

To test the T3MAPS to make sure everything is set up correctly, run `python
test_multi_column.py`. This program sends a unique pattern to each column of
the chip and then reads it back out. If this test fails, there are incorrect
connections.

Running scan viewer
---------------------

The viewer is located in the scan_analysis.py file. Execute it with

    $ python scan_analysis.py [--persist]

There will be noise on startup. If you are in persist mode, clear the noise
with x. It takes a complete cycle before the viewer resets. Note that the
history function and persistence are completely separate, so clearing the
persistence does not affect the history.


history.txt
BEGIN SCAN #0
1406916597.5
2 5 6 8 13 14 15 16 20 24 26 27 29 31 32 33 36 37 38 40 42 43 45 48 49 50 51 52 54 55 57 59 60 61 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 29 30 31 32 33 34 35 36 38 40 42 43 44 47 52 53 54 56 57 58 59 62 63 
END SCAN #0
BEGIN SCAN #1
1406916599.82
2 4 5 8 20 27 32 33 34 36 37 40 43 45 51 52 63 
0 13 36 37 39 40 53 
26 50 51 53 54 55 
16 17 23 25 26 27 28 29 33 
3 
9 35 
12 
40 41 52 54 
1 2 27 54 
11 50 

13 22 23 54 
7 8 9 
3 


0 36 
4 5 8 9 10 12 13 17 18 19 20 23 25 29 30 33 34 36 40 43 47 52 54 56 58 59 
END SCAN #1
BEGIN SCAN #2
1406916602.12
2 4 5 8 12 32 33 36 37 40 43 45 51 52 54 57 63 
6 12 45 56 
0 13 14 15 
0 1 2 12 13 14 
47 48 

27 
55 56 
54 
48 
46 

20 21 
15 16 17 


1 2 
4 5 8 9 10 11 12 13 17 18 19 20 23 25 29 33 34 36 40 43 46 47 52 54 56 58 59 
END SCAN #2
BEGIN SCAN #3
1406916604.41
0 2 4 5 8 16 20 24 32 33 36 37 40 43 45 50 51 52 63 
12 17 18 28 29 
14 33 63 

9 21 22 23 24 
9 12 13 49 


27 54 
28 


48 



0 15 16 
4 5 8 9 10 12 13 17 18 19 20 23 25 29 33 34 36 40 43 47 52 54 56 58 59 
END SCAN #3
BEGIN SCAN #4
1406916606.7
2 5 8 10 20 28 29 32 33 36 37 40 43 45 51 52 60 63 


21 
23 24 
6 45 46 53 54 
55 

33 34 54 


0 51 52 





4 5 8 9 10 12 13 17 18 19 20 23 25 29 32 33 34 36 40 43 47 52 54 56 58 59 
END SCAN #4
BEGIN SCAN #5
1406916609.0
2 5 8 16 32 33 36 37 39 40 43 45 51 52 63 
27 46 

13 
13 17 
6 55 
42 44 52 53 54 

28 50 54 59 





13 14 15 16 18 
36 
5 
4 5 6 8 9 10 11 12 13 17 18 19 20 23 25 29 33 34 36 40 43 47 52 54 56 58 59 
END SCAN #5
BEGIN SCAN #6
1406916611.3
2 5 8 16 20 27 32 33 36 37 40 43 45 49 51 52 55 56 57 60 63 
0 1 2 3 4 5 6 7 8 9 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 37 39 40 41 42 43 44 45 46 47 48 49 51 53 54 55 56 60 61 62 63 
0 2 4 6 8 10 11 12 14 18 19 21 22 24 25 29 30 32 33 35 37 40 44 45 47 49 50 51 52 53 55 56 58 60 61 62 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 30 31 32 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 14 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 32 33 34 36 37 38 39 40 42 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 
0 1 2 3 4 5 6 7 9 10 11 12 13 14 15 16 17 18 19 20 22 23 24 25 26 27 29 30 31 32 34 36 37 39 41 42 43 44 46 47 48 49 51 52 53 54 56 57 60 63 
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 39 40 42 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 

write_to_data_generator.py
"""
Module for writing an arbitrary string of bits to the data generator.

"""

import visa
import time

class DataGenerator(visa.GpibInstrument):
    """
    Facilitates communication with a DG2020A data generator.

    """

    number_of_channels = 36

    def __init__(self, port=3):
        """
        Create a new DataGenerator using the given port.
 
        """
        visa.GpibInstrument.__init__(self, "GPIB::{port}".format(port=port), board_number=0)

        self.last_write_time = time.time()
        self.wait_time = 0

    @staticmethod
    def clock(num_clocks, num_cycles_per_clock, start_on_low=True):
        """
        Get a representation of a clock to go into the data generator.

        num_clocks is the number of high/low patterns to repeat.

        num_cycles_per_clock is the number of data generator cycles that
        fits in one high (or one low), i.e. half a clock length.
 
        if start_on_low, then the clock will start on low and end on high.

        """
        low = "0" * num_cycles_per_clock
        high = "1" * num_cycles_per_clock
        cycle = low + high if start_on_low else high + low
        return cycle * num_clocks

    @staticmethod
    def blank(num_cycles):
        """
        Get a blank data matrix to go into the init_all_channels method.

        """
        return ["0"*num_cycles] * DataGenerator.number_of_channels

    def _time_sensitive(func):
        """
        A decorator to ensure we wait long enough for write
        commands to be processed.

        """
        def new_func(*args, **kwargs):
            # Sleep for the required time
            time_waited = time.time() - args[0].last_write_time
            if time_waited < args[0].wait_time:
                time.sleep(args[0].wait_time - time_waited)
            # Run the function and get the new wait time
            args[0].wait_time = func(*args, **kwargs)
            # Update the write time
            args[0].last_write_time = time.time()
            return
        return new_func
            
    @_time_sensitive            
    def set_pattern_bits(self, pattern_number, start_position, data_bits):
        """
        Write the given data bits to the data generator.

        data_bits should be a string containing 1's and 0's.
        """
        length = len(data_bits)
        length_of_length = len(str(length))
        args = dict(length=length, num_digits=length_of_length, pattern_number=pattern_number, start_position=start_position,
                    data_bits=data_bits)
        command_string_template = "data:pattern:bit {pattern_number},{start_position},{length},#{num_digits}{length}{data_bits};"
        command_string = command_string_template.format(**args)
        print "sending " + command_string
        self.write(command_string)
        return 0.002*len(data_bits)/4.0 # time to wait until can run again

    def init_channel(self, pattern_number, data_bits=None):
        """
        Re-sets the given row's pattern to the given pattern.

        If the pattern is < 64 bits long, the extra bits will be 0's.
        If no data_bits are given, simply resets the channels' pattern to 0's.

        """
        self.set_pattern_bits(pattern_number, 0, "0"*64)
        if data_bits is not None:
            self.set_pattern_bits(pattern_number, 0, data_bits)

    def init_all_channels_slow(data_generator, data_dict, num_patterns=12):
        """
        Set all the channels according to the bitstring in data_dict[pattern_num].

        If a particular pattern number is not specified, reset it to 0's.

        """
        for pattern_num in range(num_patterns):
            bit_string = data_dict.get(pattern_num, None) # default to None if data_dict[pattern_num] does not exist
            init_channel(data_generator, pattern_num, bit_string)

    def init_all_channels(self, start_address, data_matrix):
        """
        Sets all of the channels according to the bitstring in data_matrix[pattern_num].

        <start_address> is the "column" or bit number to start at (0 to 65535).
        data_matrix must be a 36-item list of strings of 1's and 0's, all of the same length.

        """
        if len(data_matrix) != DataGenerator.number_of_channels:
            raise ValueError("data_matrix must have 36 entries, 1 for each channel")

        # Set up the format of the command
        command_string_template = "data:pattern:word {address},{length},{data};"
        # Assemble the arguments address, length, and data
        args = {}
        args['address'] = start_address
        # length is the number of columns to write (i.e. length of 1 string)
        args['length'] = len(data_matrix[0])
        # data format is "#{num_digits_in_num_bytes}{num_bytes}<byte 1>...<byte n>
        data_string = "#"
        num_bytes = args['length']*5
        num_digits_in_num_bytes = len(str(num_bytes))
        data_string += str(num_digits_in_num_bytes) + str(num_bytes)
        bit_string = ""

        # each set of 5 bytes is of the form <0000><35-32><31-24>...<7-0>
        # where the numbers represent the row number (channel/pattern number)
        # and the first set of 5 bytes is for the first column, 2nd for 2nd, etc.
        try:
            for i in range(args['length']):
                bit_string += "0000"
                for pattern_num in reversed(range(len(data_matrix))): # should be range(36) always
                    bit_string += data_matrix[pattern_num][i]
        except IndexError:
            raise ValueError("all rows (strings) of data_matrix must be of the same length")
        for i in range(0, len(bit_string), 8):
            bits = bit_string[i:i+8]
            byte = chr(int(bits, 2))
            data_string += byte
        args['data'] = data_string
        command_string = command_string_template.format(**args)
        self.write(command_string)
        return 0

    def start(self):
        """
        Send the command to start outputting data.

        """
        self.write("start")

    def stop(self):
        """
        Send the command to stop outputting data.

        """
        self.write("stop")

    def update(self):
        """
        Send the command to update the data output.

        """
        self.write("data:update")

#init_all_channels_slow(data_generator, data_dict)
#init_all_channels(data_generator, 0, data_matrix)

data_dict = {3:"100000011000000110000001100000011000110010000001100000011000000110000001100000011000000100100110011100011100010001101001110100001000000110000001",5:"10"*32, 6:"0"*32+"1"*32, 9:"1"*32+"0"*32}
data_matrix = ["00001" for i in xrange(36)]
"""
init_channel(data_generator, 6, "0"*32+"1"*32)
init_channel(data_generator, 4, "01"*32)
init_channel(data_generator, 2)

# set the memory size
data_generator.write("data:msize 64;")

# check the memory size
memsize = data_generator.ask("data:msize?;")
if(memsize == ":DATA:MSIZE 64"):
    print "memory size set"
else:
    print "memory size did not set:", memsize
time.sleep(1)
data_generator.write(get_pattern_bits_command(6, 0, "0"*10))
time.sleep(1)
data_generator.write(get_pattern_bits_command(2, 0, "0"*10))
time.sleep(1)


# deletes all blocks; adds block "START" with size 64
data_generator.write("data:block:delete:all;") 
data_generator.write("data:block:rename \"UNNAMED\",\"START\";")
_command(6,0,"10"*32))
#last_write_time = time.time()
#data_generator.write("data:update;")
#print(data_generator.ask("data:pattern:bit? 6,0,10;"))


data_generator.write(init_channel(5,"101010101"))

# wait at least 0.002 seconds before writing again.
time_passed = time.time()-last_write_time
if(time_passed < 0.2):
    time.sleep(0.2-time_passed)

data_generator.write(get_pattern_bits_command(4,0,"0"*30+"0101"+"0"*30))
last_write_time = time.time()
data_generator.write("data:update;")
print(data_generator.ask("data:pattern:bit? 2,2,20;"))
"""
